cryptomite/raz.py

```diff
@@ -0,0 +1,466 @@
+    [Fore2025]_ that satisfies a given error tolerance.
+
+    Parameters
+    ----------
+    n_1 : int
+        The length of the first input (in bits).
+    k_1 : float
+        The min-entropy of the first input.
+    n_2 : int
+        The length of the second input (in bits).
+    k_2 : float
+        The min-entropy of the second input.
+    log2_error_tol : float
+        The logarithm base 2 of the acceptable extractor error.
+        Must be negative, as the extractor error is 2^log2_error.
+    m_init : int
+        The initial value for the output length (default: 1).
+    max_tests_basic : int
+        The maximum number of interations for the
+        basic parameter optimisation method, i.e. when
+        detailed_opt is set to False (default: 0).
+    max_tests_detailed : int
+        The maximum number of interations for the
+        intense parameter optimisation method, i.e.
+        when detailed_opt is set to True (default: 1000).
+    detailed_opt : bool
+        Flag to indicate the intensity of the
+        optimisation performed (default: False).
+    verbose : bool
+        If True, prints all parameters found (default: False).
+
+    Returns
+    -------
+    int :
+        The maximum output length that satisfies the
+        error tolerance.
+    """
+    max_m = 0
+    initial_tests = 100
+    ms = np.linspace(m_init, floor(k_2),
+                     min(floor(k_2) - m_init,
+                         initial_tests), dtype=int)
+    i = 0
+    STOP = 0
+    while STOP == 0:
+        m = ms[i]
+        if opt_error_raz(n_1, k_1, n_2, k_2, m,
+                         max_tests_basic=max_tests_basic,
+                         max_tests_detailed=max_tests_detailed,
+                         detailed_opt=detailed_opt,
+                         verbose=False) <= log2_error_tol:
+            max_m = m
+        else:
+            STOP = 1
+        i += 1
+        if i >= len(ms):
+            STOP = 1
+    if detailed_opt:
+        max_m += 1
+        while opt_error_raz(n_1, k_1, n_2, k_2, max_m,
+                            max_tests_basic=max_tests_basic,
+                            max_tests_detailed=max_tests_detailed,
+                            detailed_opt=detailed_opt,
+                            verbose=False) <= log2_error_tol:
+            max_m += 1
+        max_m -= 1
+    if verbose:
+        print(f'Maximum output length found: {max_m}')
+        opt_error_raz(n_1, k_1, n_2, k_2, max_m,
+                      max_tests_basic=max_tests_basic,
+                      max_tests_detailed=max_tests_detailed,
+                      detailed_opt=detailed_opt,
+                      verbose=True)
+    if max_m > n_1 / 2:
+        max_m = n_1 / 2
+    return max_m
+
+# ------ MAIN FUNCTION -------
+
+
+@staticmethod
+def from_params(
+        n_1: int,
+        k_1: float,
+        n_2: int,
+        k_2: float,
+        log2_error: float,
+        detailed_opt=False,
+        verbose: bool = True) -> Raz:
```

Comment:
Would it be better if the default for the `verbose` parameter was `False`? Just so a user of the library doesn't get stuff unexpectedly printing out unless they explicitly request it?